cpp/src/arrow/io/concurrency.h

```diff
@@ -0,0 +1,186 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <memory>
+
+#include "arrow/io/interfaces.h"
+#include "arrow/status.h"
+#include "arrow/util/checked_cast.h"
+#include "arrow/util/macros.h"
+#include "arrow/util/visibility.h"
+
+namespace arrow {
+namespace io {
+namespace internal {
+
+template <class LockType>
+class SharedLockGuard {
+ public:
+  explicit SharedLockGuard(LockType* lock) : lock_(lock) { lock_->LockShared(); }
+
+  ~SharedLockGuard() { lock_->UnlockShared(); }
+
+ protected:
+  LockType* lock_;
+};
+
+template <class LockType>
+class ExclusiveLockGuard {
+ public:
+  explicit ExclusiveLockGuard(LockType* lock) : lock_(lock) { lock_->LockExclusive(); }
+
+  ~ExclusiveLockGuard() { lock_->UnlockExclusive(); }
+
+ protected:
+  LockType* lock_;
+};
+
+// Debug concurrency checker that marks "shared" and "exclusive" code sections,
+// aborting if the concurrency rules get violated.  Does nothing in release mode.
+// Note that we intentionally use the same class declaration in debug and
+// release builds in order to avoid runtime failures when e.g. loading a
+// release-built DLL with a debug-built application, or the reverse.
+
+class ARROW_EXPORT SELockChecker {
+ public:
+  SELockChecker();
+  void LockShared();
+  void UnlockShared();
+  void LockExclusive();
+  void UnlockExclusive();
+
+  SharedLockGuard<SELockChecker> shared_guard() {
+    return SharedLockGuard<SELockChecker>(this);
+  }
+
+  ExclusiveLockGuard<SELockChecker> exclusive_guard() {
+    return ExclusiveLockGuard<SELockChecker>(this);
+  }
+
+ protected:
+  struct Impl;
+  std::shared_ptr<Impl> impl_;
+};
+
+// Concurrency wrappers for IO classes that check the correctness of
+// concurrent calls to various methods.  It is not necessary to wrap all
+// IO classes with these, only a few core classes that get used in tests.
+
+template <class Derived>
+class ARROW_EXPORT InputStreamConcurrencyWrapper : public InputStream {
+ public:
+  // TODO Close, Abort?
+  // XXX Peek() has a default implementation in InputStream
```

Comment:
Do you think there's any downside to making the inheritance virtual?

Reply:
I don't think we'd see a performance dip but `virtual` bases have poorly understood semantic overhead which we'd be passing on to implementers and users of these interfaces. In particular a reference/pointer to a virtual base must be cast using `dynamic_cast`, so anyone who writes `checked_cast<const RandomAccessFile&>(file_interface_ref)` will get errors but only when building release.

In short since this is a debug helper I'd say it's not worth introducing the complexity of a virtual base: just duplicate the implementations, they're not large.

Reply:
For reference, the only instance of public inheritance currently in our codebase is in parquet/types.cc (where it is a implementation detail internal to that translation unit).

Reply:
You mean virtual inheritance? This is false, there's actually virtual inheritance in `io/interfaces.h`, e.g.
```cpp
class ARROW_EXPORT InputStream : virtual public FileInterface, virtual public Readable {
// snip
```

But those two bases (`FileInterface`, `Readable`) are hardly ever used directly, contrary to `InputStream`.

Reply:
Woops, forgot to grep for `virtual public` as well as `public virtual`.